clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2712,34 +2705,43 @@ namespace {
       }
    }
 
-   void UpdateWidenedBounds(BoundsAnalysis &BA, const CFGBlock *Block,
-                            CheckingState &State) {
-     for (const auto item : BA.GetWidenedBounds(Block)) {
-       const VarDecl *V = item.first;
-       BoundsExpr *Bounds = item.second;
-
-       // BoundsAnalysis currently uses VarDecls as keys in the widened
-       // bounds data structure, so we create an AbstractSet for each 
-       // VarDecl in the widened bounds. TODO: use AbstractSets as keys
-       // in BoundsAnalysis (checkedc-clang issue #1015).
-       const AbstractSet *A = AbstractSetMgr.GetOrCreateAbstractSet(V);
-       auto I = State.ObservedBounds.find(A);
-       if (I != State.ObservedBounds.end())
-         I->second = Bounds;
+   void UpdateWidenedBounds(BoundsWideningAnalysis &BWA, const CFGBlock *Block,
+                            Stmt *CurrStmt, CheckingState &State) {
+
+     // Get the bounds widened before the current statement.
+     BoundsMapTy WidenedBounds = BWA.GetStmtIn(Block, CurrStmt);
+
+     // If a variable does not exist in WidenedBounds but exists in
+     // State.ObservedBounds it means that the bounds of the variable got
+     // killed before the current statement. So we reset the bounds of the
```

Comment:
This comment could be made more precise:
If a variable does not exist in `WidenedBounds` but exists in `State.ObservedBounds`, it may not be an `_Nt_array_ptr` or it is no longer a candidate for widening.